[jamie] iP

jamietan2002 opened the pull request:
# DukePro
>“Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks

- [ ] Managing deadlines (coming soon)

- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

jamietan2002 committed:
Level-1

jamietan2002 committed:
Level-2

jamietan2002 committed:
Level-3

jamietan2002 committed:
Level-4

jamietan2002 committed:
A-TextUiTesting

jamietan2002 committed:
Level-5

jamietan2002 committed:
Level-6

jamietan2002 committed:
Added level 7

jamietan2002 committed:
Updated duke.txt

jamietan2002 committed:
Added Level-8

jamietan2002 committed:
Merge branch 'branch-Level-7'
Merge Level-7

jamietan2002 committed:
Fix merge conflicts

jamietan2002 committed:
Add more OOP

jamietan2002 committed:
Add Packages

jamietan2002 committed:
Edit storage

jamietan2002 committed:
Add Gradle

jamietan2002 committed:
Add JUnit

jamietan2002 committed:
Add jar

jamietan2002 committed:
Clean up code

jamietan2002 committed:
Add JavaDocs

jamietan2002 committed:
Tweak code to follow coding standard

jamietan2002 committed:
Add find command

jamietan2002 committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-JavaDoc

jamietan2002 committed:
Merge branch 'branch-A-CodingStandard'
Merge branch-A-CodingStandard

jamietan2002 committed:
Merge branch 'branch-Level-9'
Merge Level 9

jamietan2002 committed:
Handle exceptions

jamietan2002 committed:
Fix checkstyle errors

jamietan2002 committed:
Merge branch 'A-CheckStyle'
Merge CheckStyle brach

jamietan2002 committed:
Add GUI

jamietan2002 committed:
Merge branch 'branch-Level-10'
Merge level 10

arnav-ag commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+
+    /**
```
Could this perhaps be replaced with a proper javadoc comment?

jamietan2002 replied:
Yes, will do that thank you!

arnav-ag commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+        if (input.equals("bye")) {
+            return ui.sayBye();
+        }
+        else {
```
Is this else statement needed given that the if clause returns a value?

jamietan2002 replied:
The else statement returns different values depending on the user input

arnav-ag commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+     * Replace this stub with your completed method.
+     */
+    public String getResponse(String input) {
+        Scanner sc = new Scanner(System.in);
```
This variable is never used; could this be removed?

arnav-ag commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,30 @@
+            fxmlLoader.<MainWindow>getController().setDuke(duke);
+            stage.show();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Should the user be able to see the stack trace? Maybe there's a better way to handle this error!

arnav-ag commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,121 @@
+        String first = words[0];
+
+        try {
+            if (first.equals("bye")) {
```
Perhaps the SLAP principle could be used here (using switch-case and abstracting all the logic into separate functions)?

arnav-ag commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    public void addNewTask(Task task) {
```
It would be good if javadoc could be included for all public methods!

arnav-ag commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    public ArrayList<String> findTasks(String toFind) {
```
I like how all the methods in this class are clearly abstracted properly and performing their own specific tasks (following the Single Responsibility Principle) :)

arnav-ag reviewed (commented):
LGTM, almost. Just a few nits to fix. 👍

GarethOng commented on data/tasks.txt:
```diff
@@ -0,0 +1,3 @@
+D/X/ur dad /2019-12-12
+T/ /ur mom
+T/ /urmom
```
I do not think that you intended to leave this in! 😃

GarethOng commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+
+    protected LocalDate by;
+
+    public Deadline(String description, LocalDate by) {
+        super(description);
+        this.by = by;
+    }
+    public Deadline(String description, boolean isDone, LocalDate by) {
+        super(description, isDone);
+        this.by = by;
+    }
```
Interesting usage of method overloading here!

GarethOng commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public String formatTask() {
+        return "[D] [" + super.getStatusIcon() + "] " + super.description + " (by: "
+                + by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
DateTimeFormatter.ofPattern("MMM d yyyy") can be declared as a static variable in the scenario that the similar format is used again.

clarence-chew commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+public class DukeException extends Exception {
+    /**
+     * Constructor for Duke Exception
+     * @param s custom error message when a Duke Exception is thrown
```
Do:
1. add a line between the description and the @param
2. add punctuation
```
/**
 * Constructs a Duke Exception.
 *
 * @param s Custom error message when a Duke Exception is thrown.
 */
```

https://se-education.org/guides/conventions/java/index.html#comments
"Empty line between description and parameter section"
"Punctuation behind each parameter description"

jamietan2002 replied:
ok thanks!

clarence-chew commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    public String formatTask() {
+        String var10000 = super.getStatusIcon();
```
Perhaps a better name for this variable could be `statusIcon`?

clarence-chew reviewed (commented):
Great progress, good luck for your iP!

jamietan2002 committed:
Add assertions to Parser methods

There are a few methods in Parser that will be invalid if the user
inputs are not in the correct format.

Using assertions to define assumptions about the program state so that the runtime can verify them allows bugs in the code to be detected.

Let's define assertions to ensure that assumptions in the program are
not violate.

jamietan2002 committed:
Merge pull request #1 from jamietan2002/branch-A-Assertions

Add assertions to Parser methods

jamietan2002 committed:
Improve code quality

There are redundant comments and code duplication.

Improving code quality is important to make the code more understandable
so that it can be read, understood and modified by other developers.

Lets
* remove redundant comments
* minimize code duplication
* format Javadocs to follow coding standards

jamietan2002 committed:
Merge pull request #2 from jamietan2002/branch-A-CodeQuality

Improve code quality

jamietan2002 committed:
Improve code quality

jamietan2002 committed:
Merge pull request #3 from jamietan2002/branch-A-CodeQuality

Improve code quality

jamietan2002 committed:
Add better search function

jamietan2002 committed:
Merge branch 'branch-C-BetterSearch'
Add Better Search function

jamietan2002 committed:
Improve GUI

jamietan2002 committed:
Add User Guide

jamietan2002 committed:
Fix User Guide

jamietan2002 committed:
Fix User Guide

jamietan2002 committed:
Fix storage update bug

jamietan2002 committed:
Update Javadocs
